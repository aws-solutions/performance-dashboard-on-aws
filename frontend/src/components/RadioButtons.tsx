/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";

interface Props {
    id: string;
    name: string;
    label: string;
    options: Array<{
        value: string;
        label: string;
    }>;
    register?: Function;
    required?: boolean;
    defaultValue?: string;
    hint?: string;
    error?: string;
    onChange?: Function;
    className?: string;
}

function RadioButtons(props: Props) {
    let formGroupClassName = "usa-form-group";
    if (props.error) {
        formGroupClassName += " usa-form-group--error";
    }
    if (props.className) {
        formGroupClassName += ` ${props.className}`;
    }

    const handleChange = (event: React.FormEvent<HTMLFieldSetElement>) => {
        if (props.onChange) {
            props.onChange(event);
        }
    };

    return (
        <div className={formGroupClassName}>
            <fieldset className="usa-fieldset" onChange={handleChange}>
                <legend className="usa-sr-only">{props.label}</legend>
                <span className="usa-label text-bold">{props.label}</span>
                <div className="usa-hint">{props.hint}</div>
                {props.error && (
                    <span className="usa-error-message" id="input-error-message" role="alert">
                        {props.error}
                    </span>
                )}
                <div className="margin-top-2 margin-left-2px">
                    {props.options.map((option) => {
                        const selected = props.defaultValue
                            ? props.defaultValue === option.value
                            : false;
                        return (
                            <div className="usa-radio" key={option.value}>
                                <input
                                    className="usa-radio__input"
                                    id={option.value}
                                    type="radio"
                                    defaultChecked={selected}
                                    name={props.name}
                                    value={option.value}
                                    ref={
                                        props.register &&
                                        props.register({
                                            required: props.required,
                                        })
                                    }
                                />
                                <label className="usa-radio__label" htmlFor={option.value}>
                                    {option.label}
                                </label>
                            </div>
                        );
                    })}
                </div>
            </fieldset>
        </div>
    );
}

export default RadioButtons;
