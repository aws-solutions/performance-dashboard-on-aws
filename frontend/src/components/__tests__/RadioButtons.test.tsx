/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import RadioButtons from "../RadioButtons";

test("renders radio buttons", async () => {
    const wrapper = render(
        <RadioButtons
            id="historial-figures"
            name="historical-figures"
            label="Historical Figures"
            hint="Select a historical figure"
            options={[
                {
                    label: "Harriet Tubman",
                    value: "1",
                },
                {
                    label: "Susan B. Anthony",
                    value: "2",
                },
                {
                    label: "Elizabeth Cady Stanton",
                    value: "3",
                },
            ]}
        />,
    );

    expect(wrapper.container).toMatchSnapshot();
});

test("selects a value by default", async () => {
    const { getByLabelText } = render(
        <RadioButtons
            id="historial-figures"
            name="historical-figures"
            label="Historical Figures"
            defaultValue="2"
            options={[
                {
                    label: "Harriet Tubman",
                    value: "1",
                },
                {
                    label: "Susan B. Anthony",
                    value: "2",
                },
            ]}
        />,
    );

    expect(getByLabelText("Susan B. Anthony")).toHaveAttribute("checked");
    expect(getByLabelText("Harriet Tubman")).not.toHaveAttribute("checked");
});

test("calls register function", async () => {
    const register = jest.fn();
    render(
        <RadioButtons
            id="historial-figures"
            name="historical-figures"
            label="Historical Figures"
            register={register}
            options={[
                {
                    label: "Harriet Tubman",
                    value: "1",
                },
            ]}
        />,
    );
    expect(register).toBeCalled();
});

test("renders an error", async () => {
    const wrapper = render(
        <RadioButtons
            id="historial-figures"
            name="historical-figures"
            label="Historical Figures"
            error="Something went wrong"
            options={[
                {
                    label: "Harriet Tubman",
                    value: "1",
                },
            ]}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});
