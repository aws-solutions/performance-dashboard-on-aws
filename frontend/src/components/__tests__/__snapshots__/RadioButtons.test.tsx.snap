// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an error 1`] = `
<div>
  <div
    class="usa-form-group usa-form-group--error"
  >
    <fieldset
      class="usa-fieldset"
    >
      <legend
        class="usa-sr-only"
      >
        Historical Figures
      </legend>
      <span
        class="usa-label text-bold"
      >
        Historical Figures
      </span>
      <div
        class="usa-hint"
      />
      <span
        class="usa-error-message"
        id="input-error-message"
        role="alert"
      >
        Something went wrong
      </span>
      <div
        class="margin-top-2 margin-left-2px"
      >
        <div
          class="usa-radio"
        >
          <input
            class="usa-radio__input"
            id="1"
            name="historical-figures"
            type="radio"
            value="1"
          />
          <label
            class="usa-radio__label"
            for="1"
          >
            Harriet Tubman
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`renders radio buttons 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <fieldset
      class="usa-fieldset"
    >
      <legend
        class="usa-sr-only"
      >
        Historical Figures
      </legend>
      <span
        class="usa-label text-bold"
      >
        Historical Figures
      </span>
      <div
        class="usa-hint"
      >
        Select a historical figure
      </div>
      <div
        class="margin-top-2 margin-left-2px"
      >
        <div
          class="usa-radio"
        >
          <input
            class="usa-radio__input"
            id="1"
            name="historical-figures"
            type="radio"
            value="1"
          />
          <label
            class="usa-radio__label"
            for="1"
          >
            Harriet Tubman
          </label>
        </div>
        <div
          class="usa-radio"
        >
          <input
            class="usa-radio__input"
            id="2"
            name="historical-figures"
            type="radio"
            value="2"
          />
          <label
            class="usa-radio__label"
            for="2"
          >
            Susan B. Anthony
          </label>
        </div>
        <div
          class="usa-radio"
        >
          <input
            class="usa-radio__input"
            id="3"
            name="historical-figures"
            type="radio"
            value="3"
          />
          <label
            class="usa-radio__label"
            for="3"
          >
            Elizabeth Cady Stanton
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;
